### PR TITLE
Add Encryptable::File

### PR DESCRIPTION
**Pull request by luca-bruegger**

### CODE REVIEW

**Review by mtnstar — CHANGES_REQUESTED** [review]

File: app/serializers/folder_minimal_serializer.rb [hunk 1]
```diff
@@ -4,4 +4,10 @@ class FolderMinimalSerializer < ApplicationSerializer
   attributes :id, :name, :description
 
   has_many :encryptables, serializer: EncryptableMinimalSerializer
+
+  def encryptables
+    object.encryptables.reject do |encryptable|
```
Comment (mtnstar) [inline app/serializers/folder_minimal_serializer.rb:9]: now I know what you meant ... looks like a ok solution.
problem is that all encryptable files are fetched from db this way.
but the idea is to move files anyway to same layer as credentials, ose secrets are.
so we leave it that way for now.

Comment (mtnstar) [inline app/serializers/folder_minimal_serializer.rb:9]: now I know what you meant ... looks like a ok solution.
problem is that all encryptable files are fetched from db this way.
but the idea is to move files anyway to same layer as credentials, ose secrets are.
so we leave it that way for now.

File: config/locales/en.yml [hunk 2]
```diff
@@ -437,11 +437,12 @@ en:
         user:
           new_password_invalid: Your NEW password was wrong
           old_password_invalid: Your OLD password was wrong
-        file_entry:
+        encryptable/file:
```
Comment (mtnstar) [inline config/locales/en.yml:439]: :+1:

Comment (mtnstar) [inline config/locales/en.yml:439]: :+1:

File: db/migrate/20220214135340_move_files_from_credential_to_folder.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+class MoveFilesFromCredentialToFolder < ActiveRecord::Migration[6.1]
```
Comment (mtnstar) [inline db/migrate/20220214135340_move_files_from_credential_to_folder.rb:3]: right class name?
we're not yet moving file entries to folder, are we?

Comment (mtnstar) [inline db/migrate/20220214135340_move_files_from_credential_to_folder.rb:3]: right class name?
we're not yet moving file entries to folder, are we?

File: db/migrate/20220214135340_move_files_from_credential_to_folder.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    LegacyFileEntry.new(account_id: account_id,
+                        file: file,
+                        filename: filename,
+                        description: description,
+                        content_type: content_type)
+  end
+
+  class LegacyFileEntry < ApplicationRecord
+    self.table_name = 'file_entries'
+  end
```
Comment (mtnstar) [inline db/migrate/20220214135340_move_files_from_credential_to_folder.rb:82]: well done migration :smile:

Comment (mtnstar) [inline db/migrate/20220214135340_move_files_from_credential_to_folder.rb:82]: well done migration :smile:

File: db/migrate/20220214135340_move_files_from_credential_to_folder.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+class MoveFilesFromCredentialToFolder < ActiveRecord::Migration[6.1]
+  def up
+    add_column :encryptables, :credential_id, :integer
+    add_column :encryptables, :content_type, :text
```
Comment (mtnstar) [inline db/migrate/20220214135340_move_files_from_credential_to_folder.rb:6]: content type is available as a attr ... so no need to encrypt it ?!

Comment (mtnstar) [inline db/migrate/20220214135340_move_files_from_credential_to_folder.rb:6]: content type is available as a attr ... so no need to encrypt it ?!

File: frontend/app/models/encryptable-file.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+export default class EncryptableFile extends Encryptable {
+  @attr file;
+  @belongsTo("encryptable-credential") encryptableCredential;
+
+  async save() {
+    if (this.isDeleted) {
+      return super.save();
+    }
+    const url = `/api/encryptables`;
+    const credential_id = await this.encryptableCredential.get("id");
```
Comment (mtnstar) [inline frontend/app/models/encryptable-file.js:13]: javascript uses camel case, not snake case for variables and methods
```suggestion
    const credentialId = await this.encryptableCredential.get("id");
```

Comment (mtnstar) [inline frontend/app/models/encryptable-file.js:13]: javascript uses camel case, not snake case for variables and methods
```suggestion
    const credentialId = await this.encryptableCredential.get("id");
```

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 7]
```diff
@@ -98,6 +98,40 @@
...
+      file_attributes = %w[name description]
+      expect_json_object_includes_keys(files_json_attributes, file_attributes)
+    end
+
+    it 'does not return encryptable file without access' do
+      login_as(:alice)
+
+      file = create_file
+
+      file_params = {
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:127]: is this variable required? id: file.id looks pretty short ...

Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:127]: is this variable required? id: file.id looks pretty short ...

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 8]
```diff
@@ -575,29 +610,97 @@
       expect(response).to have_http_status(201)
     end
 
-    context 'DELETE destroy' do
-      it 'cant destroy an encryptable if not in team' do
-        login_as(:alice)
+    it 'creates new encryptable file if part of the team' do
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:613]:
```suggestion
    it 'creates new encryptable file' do
```

Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:613]:
```suggestion
    it 'creates new encryptable file' do
```

File: spec/system/encryptable_files_modal_system_spec.rb [hunk 9]
```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe 'Encryptable Files Modal', type: :system, js: true do
+  # TODO: adjust tests after implementing/fixing encryptable files modal
```
Comment (mtnstar) [inline spec/system/encryptable_files_modal_system_spec.rb:6]: TODO ? do we have to fix this?

Comment (mtnstar) [inline spec/system/encryptable_files_modal_system_spec.rb:6]: TODO ? do we have to fix this?

**Review by mtnstar — CHANGES_REQUESTED** [review]
thanks for the great effort ! huge PR ! lot's of code ... almost there ...

* check file content type column
* find an elegant solution for determine sti file type in controller by json

File: app/controllers/api/encryptables_controller.rb [hunk 10]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 class Api::EncryptablesController < ApiController
+  include Encryptables
+
   self.permitted_attrs = [:name, :description, :folder_id, :tag]
```
[the remark this answers is not in the document]
Comment (luca-bruegger) [reply app/controllers/api/encryptables_controller.rb:6]: This already happens within the `permitted_attrs` method.

File: app/controllers/api/encryptables_controller.rb [hunk 11]
```diff
@@ -50,44 +53,50 @@ def update
...
+  # rubocop:disable Metrics/MethodLength
   def model_class
-    if action_name == 'create' &&
-       params.dig('data', 'attributes', 'type') == 'ose_secret'
+    if create_ose_secret?
       Encryptable::OseSecret
     elsif action_name == 'destroy'
       Encryptable
     elsif @encryptable.present?
       encryptable.class
```
[the remark this answers is not in the document]
Comment (luca-bruegger) [reply app/controllers/api/encryptables_controller.rb:65]: This is going to be resolved on the follow-up branch: https://github.com/puzzle/cryptopus/tree/feature/530-encryptable-file-controller-cleanup

File: app/controllers/api/encryptables_controller.rb [hunk 12]
```diff
@@ -16,19 +18,20 @@ def index(options = {})
...
-    authorize encryptable
-    encryptable.decrypt(decrypted_team_password(team))
+    authorize entry
+    entry.decrypt(decrypted_team_password(team))
     render_entry
   end
 
   # POST /api/encryptables
-  def create
+  def create(options = {})
```
[the remark this answers is not in the document]
Comment (mtnstar) [reply app/controllers/api/encryptables_controller.rb:28]: remove this options ... use before_save from dry crud as discussed

**Review by mtnstar — APPROVED** [review]
:partying_face:

**Pull request merged**